sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/RewriteDeleteFromTable.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.expressions.{EqualNullSafe, Expression, Literal, Not}
+import org.apache.spark.sql.catalyst.plans.logical.{DeleteFromTable, Filter, LogicalPlan, ReplaceData}
+import org.apache.spark.sql.connector.catalog.{SupportsDelete, SupportsRowLevelOperations}
+import org.apache.spark.sql.connector.write.RowLevelOperation.Command.DELETE
+import org.apache.spark.sql.connector.write.RowLevelOperationTable
+import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
+
+/**
+ * Assigns a rewrite plan for v2 tables that support rewriting data to handle DELETE statements.
+ *
+ * If a table implements [[SupportsDelete]] and [[SupportsRowLevelOperations]], this rule assigns
+ * a rewrite plan but the optimizer will check whether this particular DELETE statement can be
+ * handled by simply passing delete filters to the connector. If yes, the optimizer will then
+ * discard the rewrite plan.
```

Comment:
Where the optimizer will do the check and discard the rewrite plan? In `ReplaceRewrittenRowLevelCommands`, seems it simply gets the rewrite plan of the `RowLevelCommand` (e.g. `DeleteFromTable`).

Reply:
It is done in `OptimizeMetadataOnlyDeleteFromTable`. That rule checks if we can handle this delete by just passing data source filters. If yes, the rewrite plan is nullified.

That being said, we will call `SupportsDelete$canDeleteWhere` two times now. The first call will be in the optimizer and  the second one will be in `DataSourceV2Strategy`.

Reply:
Instead of having `ReplaceRewrittenRowLevelCommands` and `OptimizeMetadataOnlyDeleteFromTable` in the optimizer, we can add more logic to `DataSourceV2Strategy`. Let me try that.

Reply:
Well, I remember why I did this as an optimizer rule. It is because I wanted to avoid job planning when DELETE can be handled using filters.

Reply:
I added `DeleteFromTableWithFilters` and we call `SupportsDelete$canDeleteWhere` only once now.

Reply:
This thread is a bit confusing now. I'll resolve it. There are a few others where we discuss the same thing.